### PR TITLE
feat: custom fields support [CU-28t5nw1] [CU-2fjq0m6]

### PR DESCRIPTION
**Pull request by ltsNotMike**
## Ticket

N/A

## Summary

- Support for managing custom fields
- Support for 2 types of custom fields (checkbox and text)
- Support for editing custom fields in NavigationItem Form
- Refactors

## Test Plan

- [x] Create a project with a navigation plugin installed
- [x] Add new custom fields through the Settings Page
- [x] Try adding fields with incorrect data (empty fields or repeated name fields)
- [x] Edit/ Remove custom field and check if navigation flow still works fine
- [x] Add some items to navigation and fill its custom fields
- [x] Check if fields are being rendered through API and graphql
- [x] Try enabling/disabling the audience field and check if the navigation flow works fine
- [x] Try adding audience fields to navigation items and check if it's rendered both in API and in graphql

### CODE REVIEW

**Comment by cyp3rius** [comment]
Task linked: [CU-28t5nw1 Extendable additional fields](https://app.clickup.com/t/28t5nw1)

File: admin/src/pages/SettingsPage/types.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// TODO: This type should be replaced with new version of content type in new strapi-typed
+export type ContentTypeToFix = StrapiContentTypeFullSchema & { available: boolean, isSingle: boolean, plugin: string }
+
+export type IPreparePayload = (payload: { form: RawPayload, pruneObsoleteI18nNavigations: boolean }) => NavigationPluginConfig;
+export type IOnSave = (form: RawPayload) => void;
+export type IOnPopupClose = (isConfirmed: boolean) => void;
+export type IHandleSetContentTypeExpanded = (key: string | undefined) => void;
+export type IPrepareNameFieldFor = (uid: string, current: Record<string, string[] | undefined>, value: string[]) => Record<string, string[] | undefined>;
+export type RestartReasons = 'I18N' | 'GRAPH_QL' | 'I18N_NAVIGATIONS_PRUNE';
+export type RestartStatus = { required: boolean, reasons?: RestartReasons[] }
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/types.ts:21]:
```suggestion
export type RestartStatus = { required: true, reasons?: RestartReasons[] } | { required: false }
```

This we can eliminate illogical states like no restart required with reasons.

File: admin/src/pages/SettingsPage/types.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  audienceFieldChecked: boolean;
+  allowedLevels: number;
+  populate: Record<string, string[]>;
+  i18nEnabled: boolean;
+}
+
+// TODO: This type should be replaced with new version of content type in new strapi-typed
+export type ContentTypeToFix = StrapiContentTypeFullSchema & { available: boolean, isSingle: boolean, plugin: string }
+
+export type IPreparePayload = (payload: { form: RawPayload, pruneObsoleteI18nNavigations: boolean }) => NavigationPluginConfig;
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/types.ts:15]: `I` in a name stands for `Interface`. I'd rather to see `export interface` :D and leave name "as is"

File: admin/src/pages/SettingsPage/CustomFieldTable/index.tsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        onConfirm={handleConfirm}
+        onCancel={handleCancel}
+      />
+      <Table
+        colCount={4}
+        rowCount={data.length + 1}
+        footer={
+          <TFooter
+            onClick={(e: React.FormEvent) => { e.preventDefault(); onOpenModal(null); }}
+            icon={<Plus />}
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/CustomFieldTable/index.tsx:84]: This icon can be be also extracted like `refreshIcon`

File: server/services/client.ts [hunk 4]
```diff
@@ -333,12 +332,13 @@ const clientService: (context: StrapiContext) => IClientService = ({ strapi }) =
         default:
           const publishedItems = items.filter(filterOutUnpublished);
           const result = isNil(rootPath) ? items : filterByPath(publishedItems, rootPath).items;
-          return result.map((item: NavigationItemEntity<ContentTypeEntity>) => ({
+          return result.map(({additionalFields, ...item}: NavigationItemEntity<ContentTypeEntity>) => ({
```
[the remark this answers is not in the document]
Comment (ltsNotMike) [reply server/services/client.ts:335]: We definitely should but this one is not `spread` but `rest` operator.

[the remark this answers is not in the document]
Comment (CodeVoyager) [reply server/services/client.ts:335]: Right, silly me 😅

File: admin/src/pages/SettingsPage/index.tsx [hunk 5]
```diff
@@ -167,16 +195,35 @@ const SettingsPage = () => {
...
   const allowedLevels = navigationConfigData?.allowedLevels || 2;
   const nameFields = navigationConfigData?.contentTypesNameFields || {}
   const populate = navigationConfigData?.contentTypesPopulate || {}
   const i18nEnabled = navigationConfigData?.i18nEnabled ?? false
   const isI18NPluginEnabled = navigationConfigData?.isI18NPluginEnabled;
   const defaultLocale = navigationConfigData?.defaultLocale;
 
+  const handleSubmitCustomField = (field: NavigationItemCustomField) => {
+    const filteredFields = customFields.filter(f => f.name !== field.name);
+    setCustomFields([...filteredFields, field]);
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/index.tsx:210]:
```suggestion
    const fields = customFields.map(current => current.name === field.name ? field : current);
    setCustomFields(fields);
```

File: admin/src/pages/SettingsPage/types.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  i18nEnabled: boolean;
+}
+
+// TODO: This type should be replaced with new version of content type in new strapi-typed
+export type ContentTypeToFix = StrapiContentTypeFullSchema & { available: boolean, isSingle: boolean, plugin: string }
+
+export type PreparePayload = (payload: { form: RawPayload, pruneObsoleteI18nNavigations: boolean }) => NavigationPluginConfig;
+export type OnSave = (form: RawPayload) => void;
+export type OnPopupClose = (isConfirmed: boolean) => void;
+export type HandleSetContentTypeExpanded = (key: string | undefined) => void;
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/types.ts:18]: `ChangeEffect` maybe?

File: admin/src/translations/en.json [hunk 7]
```diff
@@ -113,7 +115,26 @@
...
+	"pages.settings.form.customFields.popup.header.edit": "Edit custom field",
+	"pages.settings.form.customFields.popup.header.new": "Add new custom field",
+  "pages.settings.form.customFields.popup.name.label": "Custom field name",
+  "pages.settings.form.customFields.popup.name.placeholder": "example_name",
+  "pages.settings.form.customFields.popup.name.description": "Name of the custom field must be unique", 
+  "pages.settings.form.customFields.popup.label.label": "Custom field label",
+  "pages.settings.form.customFields.popup.label.placeholder": "Example label",
+  "pages.settings.form.customFields.popup.label.description": "This label will be displayed on navigation item form",
+  "pages.settings.form.customFields.popup.type.label": "Custom field type",
+  "components.navigationItem.action.newItem": "Add nested item",
```
Comment (CodeVoyager) [inline admin/src/translations/en.json:137]: Wrong indentation

**Review by CodeVoyager — APPROVED** [review]

File: admin/src/pages/SettingsPage/index.tsx (hunk 5 above)
Comment (ltsNotMike) [reply admin/src/pages/SettingsPage/index.tsx:210]: At first, it looked like a great way to make this part cleaner. But after testing I realized why I hadn't done it this way in the first place. Solution with filter works for both situations when the name exists in `customFields` and when it does not. To make it simpler I could divide it into two separate methods:`handleNewCustomField` and `handleEditCustomField`.

**Review by CodeVoyager — APPROVED** [review]

File: __mocks__/strapi.ts [hunk 8]
```diff
@@ -143,9 +146,10 @@ const plugins = (strapi: IStrapi): StringMap<StrapiPlugin> => ({
     config: (key: string) => ({
       ...defaultConfig.default,
       contentTypes: ['api::pages.pages'],
+      additionalFields: [{ name: 'string_test_field', label: "Test field", type: 'string', enabled: true }]
     })[key],
     get controllers() { return {} },
-    controller() { return () => {}  },
+    controller(): ToBeFixed { return {} },
```
Comment (CodeVoyager) [inline __mocks__/strapi.ts:152]: Current version does not require hacking ;)

File: admin/src/components/AdditionalFieldInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+}) => {
+  const toggleNotification = useNotification();
+  const { formatMessage } = useIntl();
+  const defaultInputProps = useMemo(() => ({
+    id: field.name,
+    name: field.name,
+    label: field.label,
+    disabled: isLoading,
+    error: error && formatMessage(error),
+
```
Comment (CodeVoyager) [inline admin/src/components/AdditionalFieldInput/index.tsx:38]: not needed space

File: admin/src/components/AdditionalFieldInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      return (
+        <Select
+          {...defaultInputProps}
+          onChange={(v: string) => onChange(field.name, v)}
+          value={isNil(value) ? field.multi ? [] : null : value}
+          multi={field.multi}
+          withTags={field.multi}
+        >
+          {field.options.map((option, index) => (
+            <Option key={index} value={option}>
```
Comment (CodeVoyager) [inline admin/src/components/AdditionalFieldInput/index.tsx:76]: Use compound `key`, please

File: admin/src/components/TextArrayInput/index.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  id?: string;
+  name?: string;
+  label?: string;
+  disabled?: boolean;
+  // TODO: [ @ltsNotMike ] Fix this before commit 
+  error?: ToBeFixed;
+}
+
+const TextArrayInput: React.FC<IProps> = ({ onChange, initialValue, ...props}) => {
+  const [value, setValue] = useState(!!initialValue ? initialValue.reduce((acc, cur) => acc+cur+"; ", "") : "");
```
Comment (CodeVoyager) [inline admin/src/components/TextArrayInput/index.tsx:18]:
```suggestion
  const [value, setValue] = useState(Array.isArray(initialValue) ? initialValue.reduce((acc, cur) => [acc, cur, "; "].join(""), "") : "");
```

or alternatively you can use template string

File: admin/src/components/TextArrayInput/index.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  const handleOnChange = (e: React.BaseSyntheticEvent) => {
+    const newValue: string = e.target.value;
+    const valuesArray = newValue
+      .split(';')
+      .map(v => v.trim())
+      .filter(v => !!v.length);
+    setValue(e.target.value)
+    onChange(valuesArray);
+  }
```
Comment (CodeVoyager) [inline admin/src/components/TextArrayInput/index.tsx:27]:
```suggestion
  const handleOnChange = ({ target: { value } }: React.BaseSyntheticEvent) => {
    const valuesArray = value
      .split(';')
      .map(v => v.trim())
      .filter(v => v.length);

    setValue(value)
    onChange(valuesArray);
  }
```

File: admin/src/hooks/useAllContentTypes.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+//@ts-ignore
+import { useQuery } from 'react-query';
+import { fetchAllContentTypes } from '../utils';
+
+const useAllContentTypes = () => {
+  const { isLoading, data, error } = useQuery('contentTypes', () =>
+    fetchAllContentTypes()
+  );
+  return { data, isLoading, error };
+};
```
Comment (CodeVoyager) [inline admin/src/hooks/useAllContentTypes.ts:10]:
```suggestion
import { pick } from 'lodash';
import { fetchAllContentTypes } from '../utils';

const useAllContentTypes = () => 
  pick(
      useQuery('contentTypes', fetchAllContentTypes),
      [ "data", "isLoading", "error" ],
  );
```

File: admin/src/pages/SettingsPage/components/CustomFieldTable/index.tsx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+          id: getTradId(`${tradPrefix}confirmation.error`),
+          default: 'Something went wrong',
+        }
+      });
+    } else {
+      onRemoveCustomField(fieldToRemove);
+    }
+
+    cleanup();
+  }
```
Comment (CodeVoyager) [inline admin/src/pages/SettingsPage/components/CustomFieldTable/index.tsx:66]: `useCallback`, please

File: admin/src/pages/View/components/NavigationManager/Form/index.tsx [hunk 15]
```diff
@@ -24,6 +25,10 @@ export const Form = ({
   isLoading,
   validationSchema,
 }: Props) => {
+  const initialValues: Navigation= {
+    id: get(navigation, "id", ""),
+    name: get(navigation, "name", ""),
+  }
```
Comment (CodeVoyager) [inline admin/src/pages/View/components/NavigationManager/Form/index.tsx:31]: Why this change happened?

File: admin/src/translations/en.json [hunk 16]
```diff
@@ -79,10 +79,12 @@
 	"notification.navigation.item.relation.status.draft": "draft",
 	"notification.navigation.item.relation.status.published": "published",
 	"notification.error": "Error while processing request.",
+  "notification.error.customField.type": "Unsupported type of custom field",
```
Comment (CodeVoyager) [inline admin/src/translations/en.json:82]: looks like indentation issue(s)

File: admin/src/utils/api.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+export const updateNavigationConfig = ({ body }: { body: NavigationPluginConfig }) =>
+  request(`/${pluginId}/config`, { method: 'PUT', body }, true);
+
+export const restoreNavigationConfig = () =>
+  request(`/${pluginId}/config`, { method: 'DELETE' }, true);
+
+export const fetchAllContentTypes = async () => {
+  const { data } = await request('/content-manager/content-types', { method: 'GET' });
+  return data;
+}
```
Comment (CodeVoyager) [inline admin/src/utils/api.ts:18]:
```suggestion
export const fetchAllContentTypes = () => 
  request('/content-manager/content-types', { method: 'GET' }).then(prop("data"));
```

File: admin/src/utils/enums.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    LOADING: 'LOADING',
+    ERROR: 'ERROR',
+};
+
+export const resolvedResourceFor = (value: unknown) => ({
+    type: ResourceState.RESOLVED,
+    value,
+});
+
+export const errorStatusResourceFor = (errors: unknown) => ({
```
Comment (CodeVoyager) [inline admin/src/utils/enums.ts:26]:
```suggestion
export const errorStatusResourceFor = (errors: Array<Error>) => ({
```

File: admin/src/utils/enums.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  NAVIGATION_ITEM: 'navigationItem'
+};
+
+export const ResourceState = {
+    RESOLVED: 'RESOLVED',
+    LOADING: 'LOADING',
+    ERROR: 'ERROR',
+};
+
+export const resolvedResourceFor = (value: unknown) => ({
```
Comment (CodeVoyager) [inline admin/src/utils/enums.ts:21]:
```suggestion
export const resolvedResourceFor = <T = unknown>(value: T) => ({
```

File: server/graphql/types/navigation-item.js [hunk 20]
```diff
@@ -1,29 +1,55 @@
...
+                t.boolean(field.name);
+                break;
+              case 'select':
+                if (field.multi)
+                  t.list.string(field.name);
+                else
+                  t.string(field.name);
+                break;
+              default:
+                throw new Error(`Type "${field.type}" is unsupported by custom fields`);
```
Comment (CodeVoyager) [inline server/graphql/types/navigation-item.js:47]:
```suggestion
                throw new Error(`Type "${JSON.stringify(field.type)}" is unsupported by custom fields`);
```

If `field.type` happens to be something beyond primitive value

File: server/services/admin.ts [hunk 21]
```diff
@@ -231,7 +254,12 @@ const adminService: (context: StrapiContext) => IAdminService = ({ strapi }) =>
   async updateConfig(newConfig: NavigationPluginConfig): Promise<void> {
     const commonService = getPluginService<ICommonService>('common');
     const pluginStore = await commonService.getPluginStore()
+    const config = await pluginStore.get<string, NavigationPluginConfig>({ key: 'config' });
+    validateAdditionalFields(newConfig.additionalFields);
     await pluginStore.set({ key: 'config', value: newConfig });
+
+    const removedFields = differenceBy(config.additionalFields, newConfig.additionalFields, 'name').filter(i => i !== 'audience') as NavigationItemCustomField[];
+    !isEmpty(removedFields) && await commonService.pruneCustomFields(removedFields);
```
Comment (CodeVoyager) [inline server/services/admin.ts:262]: `if` is more readable imho

File: types/utils.ts [hunk 22]
```diff
@@ -11,6 +11,8 @@ export type RenderType = keyof RenderTypeOptions;
 export type StrapiRoutesTypes = 'admin' | 'content-api';
 export type ToBeFixed = any;
 export type DateString = string;
+export type ChangeEffect<Args> = (args: Args) => void; 
+export type VoidEffect = ChangeEffect<void>;
```
Comment (CodeVoyager) [inline types/utils.ts:15]: Keep in mind that I reimplemented `Effect` in `master`, please.

File: admin/src/pages/View/components/NavigationManager/Form/index.tsx (hunk 15 above)
Comment (ltsNotMike) [reply admin/src/pages/View/components/NavigationManager/Form/index.tsx:31]: Types mismatched. Props define `navigation` as `Partial<Navigation>` and `onChangeBase` as `(n: Navigation)=> void`. This creates conflict inside Formik and prints out error for initialValues.

File: admin/src/translations/en.json (hunk 16 above)
Comment (ltsNotMike) [reply admin/src/translations/en.json:82]: Oh, man... I need a better auto-formatting solution 🙄

File: admin/src/pages/View/components/NavigationManager/Form/index.tsx (hunk 15 above)
Comment (CodeVoyager) [reply admin/src/pages/View/components/NavigationManager/Form/index.tsx:31]: makes sense

**Comment by ltsNotMike** [comment]
One more commit is waiting to be added but is blocked by [this](https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/pull/243).

**Review by cyp3rius — APPROVED** [review]
